integration-tests/tests/client/sharding_upgrade.rs

```diff
@@ -263,12 +326,139 @@ fn test_shard_layout_upgrade_simple() {
+}
+
+const GAS_1: u64 = 300_000_000_000_000;
+const GAS_2: u64 = GAS_1 / 3;
+
+fn gen_cross_contract_transactions(
+    signer: &InMemorySigner,
+    receiver: &AccountId,
+    nonce: u64,
+    block_hash: &CryptoHash,
+) -> SignedTransaction {
+    let data = serde_json::json!([
+        {"create": {
+        "account_id": "test2",
+        "method_name": "call_promise",
+        "arguments": [],
+        "amount": "0",
+        "gas": GAS_2,
+        }, "id": 0 },
+        {"then": {
+        "promise_index": 0,
+        "account_id": "test3",
+        "method_name": "call_promise",
+        "arguments": [],
+        "amount": "0",
+        "gas": GAS_2,
+        }, "id": 1}
+    ]);
+
+    SignedTransaction::from_actions(
+        nonce,
+        signer.account_id.clone(),
+        receiver.clone(),
+        signer,
+        vec![Action::FunctionCall(FunctionCallAction {
+            method_name: "call_promise".to_string(),
+            args: serde_json::to_vec(&data).unwrap(),
+            gas: GAS_1,
+            deposit: 0,
+        })],
+        block_hash.clone(),
+    )
+}
+
+// Test cross contract calls
+// This test case tests postponed receipts and delayed receipts
+#[test]
+fn test_shard_layout_upgrade_cross_contract_calls() {
+    init_test_logger();
+
+    // setup
+    let epoch_length = 5;
+    let mut test_env = TestShardUpgradeEnv::new(epoch_length, 4, 4, 100, Some(100_000_000_000_000));
+
+    let genesis_hash = test_env.env.clients[0].chain.genesis_block().hash().clone();
+    test_env.set_init_tx(
+        test_env.initial_accounts[0..test_env.num_validators]
+            .iter()
+            .map(|account_id| {
+                let signer = InMemorySigner::from_seed(
+                    account_id.clone(),
+                    KeyType::ED25519,
+                    &account_id.to_string(),
+                );
+                SignedTransaction::from_actions(
+                    1,
+                    account_id.clone(),
+                    account_id.clone(),
+                    &signer,
+                    vec![Action::DeployContract(DeployContractAction {
+                        code: near_test_contracts::rs_contract().to_vec(),
+                    })],
+                    genesis_hash.clone(),
+                )
+            })
+            .collect(),
     );
+
+    let mut nonce = 100;
+    let generate_txs: &mut dyn FnMut(usize) -> Vec<SignedTransaction> =
+        &mut |size: usize| -> Vec<SignedTransaction> {
+            let signer0 =
+                InMemorySigner::from_seed("test0".parse().unwrap(), KeyType::ED25519, "test0");
+            let mut txs = vec![];
+            for _ in 0..size {
+                txs.push(gen_cross_contract_transactions(
+                    &signer0,
+                    &"test1".parse().unwrap(),
+                    nonce,
+                    &genesis_hash,
+                ));
+                nonce += 1;
+            }
+            txs
```

Comment:
Alternative, not necssary better way is something like `std::iter::repeat_with(|| make_tx).take(size).collect()`